canals/pipeline/connections.py

```diff
@@ -23,6 +23,42 @@ def parse_connection_name(connection: str) -> Tuple[str, Optional[str]]:
     return connection, None
 
 
+def _type_is_compatible(source_type, dest_type):
+    """
+    Checks whether the source type is equal or a subtype of the destination type. Used to validate pipeline connections.
+
+    Note: this method has no pretense to perform proper type matching. It especially does not deal with aliasing of
+    typing classes such as `List` or `Dict` to their runtime counterparts `list` and `dict`. It also does not deal well
+    with "bare" types, so `List` is treated differently from `List[Any]`, even though they should be the same.
+
+    Consider simplifying the typing of your components if you observe unexpected errors during component connection.
+    """
+    if source_type == dest_type or dest_type is Any:
+        return True
+
+    if source_type is Any:
+        return False
+
+    try:
+        if issubclass(source_type, dest_type):
+            return True
+    except TypeError:  # typing classes can't be used with issubclass, so we deal with them below
+        pass
+
+    source_origin = get_origin(source_type)
+    dest_origin = get_origin(dest_type)
+
+    if source_origin is not Union and dest_origin is Union:
+        return any(_type_is_compatible(source_type, union_arg) for union_arg in get_args(dest_type))
+
+    source_args = get_args(source_type)
+    dest_args = get_args(dest_type)
+    if not source_origin or not dest_origin or source_origin != dest_origin or len(source_args) > len(dest_args):
+        return False
```

Comment:
I think it's less confusing if we split this check.

```suggestion
    if not source_origin or not dest_origin or source_origin != dest_origin:
        return False

    source_args = get_args(source_type)
    dest_args = get_args(dest_type)
    if len(source_args) > len(dest_args):
        return False
```